aoc2019/day07/amplifiers.py

```diff
@@ -14,41 +16,37 @@ def find_largest_output(self):
             output = 0
             for index, amplifier in enumerate(self.amplifiers):
                 phase_setting = combination[index]
-                amplifier.reinitialize()
-                output = amplifier.run_program([phase_setting, output])[0]
+                program = amplifier.load_program()
+                program.send(None)
+                program.send(phase_setting)
+                output = program.send(output)[0]
             if not largest_output or output > largest_output:
                 largest_output = output
         return largest_output
 
-    def is_halted(self, amplifier):
-        return amplifier.registers[amplifier.instruction_pointer] == 99
-
-    def initialize_amplifiers(self, combination):
-        previous_result = [0]
-        for index, phase in enumerate(combination):
-            previous_result.insert(0, phase)
-            try:
-                previous_result = self.amplifiers[index].run_program(previous_result)
-            except MissingInputError:
-                previous_result = self.amplifiers[index].output
-                self.amplifiers[index].output = []
-        return previous_result
+    def initialize_programs(self, combination):
+        programs = []
+        for index, amplifier in enumerate(self.amplifiers):
+            next_program = amplifier.load_program()
+            next_program.send(None)
+            next_program.send(combination[index])
+            programs.append(next_program)
+        return programs
 
     def run_feedback_loop(self, combination):
-        self.reinitialize_amplifiers()
-        previous_result = self.initialize_amplifiers(combination)
-        while not self.is_halted(self.amplifiers[-1]):
-            for amplifier in self.amplifiers:
-                try:
-                    previous_result = amplifier.run_program(previous_result)
-                except MissingInputError:
-                    previous_result = amplifier.output
-                    amplifier.output = []
-        return previous_result[0]
-
-    def reinitialize_amplifiers(self):
-        for amplifier in self.amplifiers:
-            amplifier.reinitialize()
+        programs = self.initialize_programs(combination)
+        results = defaultdict(lambda: [0])
+        while True:
```

Comment:
Can also make a has_next function in int_code to try and get the next value and return false otherwise. This may be complex because send will require some form of input (1 arbitrary number) in cases that care about has_next. I'm inclined to leave it be, since this is the only case we need to check to see we've halted instead of providing the correct number of inputs